### PR TITLE
Typehinting

### LINKED ISSUE #68: Add support for type hints
A straight forward way to improve the code quality is to enable runtime support for type hints via `typing`.

### PR DESCRIPTION
**Pull request by wiseodd**
Closes #68 

WIP and blocked by #144

### CODE REVIEW

**Comment by wiseodd** [comment]
Ready for review!

Typehinting and then typechecking with `pyright` really shows that there are many ambiguous typing going on in our code. To be fair due to Python allowing it. Good idea to be strict with typing for next projects lol---lesson learned.

File: laplace/curvature/asdfghjkl.py [hunk 1]
```diff
@@ -101,36 +113,47 @@ def _get_kron_factors(self, curv, M):
...
+    ) -> tuple[torch.Tensor, Kron]:
         with torch.no_grad():
             if self.last_layer:
-                f, X = self.model.forward_with_features(X)
+                f, x = self.model.forward_with_features(x)
             else:
-                f = self.model(X)
+                f = self.model(x)
             loss = self.lossfunc(f, y)
         curv = fisher_for_cross_entropy(
```
Comment (wiseodd) [inline laplace/curvature/asdfghjkl.py:155]: The only thing that's still missing is the type for `curv` here. @aleximmer @runame do you know what's its type? I dug into ASDL v0.1's code and came back even more confused...

**Comment by wiseodd** [comment]
@runame ready for review!

File: laplace/curvature/asdfghjkl.py (hunk 1 above)
Comment (runame) [reply laplace/curvature/asdfghjkl.py:164]: This function does not exist in ASDL, only in its predecessor `asdfghjkl`. The return type will be `FisherExactCrossEntropy`, `FisherMCCrossEntropy`, or `FisherEmpCrossEntropy` depending on `self._ggn_type`. You should be able to either import these classes from `asdfghjkl` directly or from `asdfghjkl.fisher`.

**Review by runame — APPROVED** [review]
Didn't carefully check every diff, but LGTM. I think this definitely improved the code quality!

**Pull request merged**